### ISSUE TITLE
[RFE] Migrate Microsoft Graph API client from manicminer/hamilton to microsoftgraph/msgraph-sdk-go

### ISSUE THREAD

**Issue by JonCrowther** [rancher/rancher#45009]
**Is your feature request related to a problem? Please describe.**
Currently to interact with the Microsoft Graph API, Rancher uses this 3rd party [Hamilton library](https://github.com/manicminer/hamilton). There are some security, performance and maintenance concerns. Additionally, past issues we've had with the Hamilton library have taken a while to get addressed.

**Describe the solution you'd like**
We would like to move to the officially supported [Microsoft go-sdk](https://github.com/microsoftgraph/msgraph-sdk-go).

**Describe alternatives you've considered**
We could continue to use the existing 3rd party library, and open issues in that repo as needed. Historically that has worked for smaller changes, but if there's a significant change/feature we need, it can be very slow to get those changes into the Hamilton library.

**Additional context**
There is ongoing work for our Azure AD support, so the earlier we are able to migrate, the less challenging the migration will be.

**Comment by bigkevmcd** [comment]
# Migrating to the Azure SDK for Go

Using https://github.com/microsoftgraph/msgraph-sdk-go

Rancher currently has **two** clients for Azure, the now [deprecated](https://github.com/rancher/rancher/blob/26fbb330c6f01af2c07d4d60b7692333a62c5fbf/pkg/auth/providers/azure/clients/client.go#L26-L42) `AzureADGraphClient` and the `MSGraphClient`.

I don't believe we know how many users use the deprecated client, and I can't find the deprecation timeline for it.

## Azure SDK AzureClient

We would need to implement [`AzureClient`](https://github.com/rancher/rancher/blob/26fbb330c6f01af2c07d4d60b7692333a62c5fbf/pkg/auth/providers/azure/clients/client.go#L14-L24) but using the Microsoft SDK client.

From a technical PoV this is simply going through the code in the current [MS Graph Client](https://github.com/rancher/rancher/blob/26fbb330c6f01af2c07d4d60b7692333a62c5fbf/pkg/auth/providers/azure/clients/ms_graph_client.go#L31) looking up the methods in the msgraph-sdk [GoDoc](https://pkg.go.dev/github.com/microsoftgraph/msgraph-sdk-go#GraphBaseServiceClient.Users) and translating the calls that we use (based on the hamilton package) into calls to the Microsoft package.

It's complicated by the documentation for the package, which is not [particularly](https://pkg.go.dev/github.com/microsoftgraph/msgraph-sdk-go@v1.37.0/users) helpful.

There are some useful examples in the package [documentation](https://github.com/microsoftgraph/msgraph-sdk-go?tab=readme-ov-file#4-getting-results-that-span-across-multiple-pages) and [snippets](https://github.com/microsoftgraph/microsoft-graph-docs-contrib/blob/f166be1128c4b87d0eba2e7c402518cd23c40b12/includes/snippets/go/v1/customsecurityattribute-get-for-user-go-snippets.md).

## Before reimplementing the client

 * Do we want to fully deprecate the _old_ client? Can we?
 * Does this replace the _new_ client?
 * Do we want more than one _deprecated_ client?

The `AzureClient` interface has eight methods, and we have two existing implementations, so it shouldn't be particularly hard to migrate, we need to figure out how we deal with the deprecation elements.

**Comment by bigkevmcd** [comment]
## Validation Template

## What was fixed, or what change has occurred

We removed manicminer/hamilton and replaced it with calls to the underlying MS Azure library.

## Areas or cases that should be tested

Azure AD authentication.

 * Login
 * Search for users and groups

## What areas could experience regressions

Azure AD authentication and user/groups searching

## Are the repro steps accurate/minimal?

**Comment by anupama2501** [comment]
Re opening for additional validations

**Comment by anupama2501** [comment]
Verified the following cases on v2.9-head `637346f` and on an upgrade from v2.8.5 to v2.9-head `637346f`

- From Users and authentication, enable an auth provider depending on the requirement.
- Once enabled, from the local cluster explorer, navigate to secrets >> Select the namespace cattle-global-data.
    - Verify the secrets for the enabled auth provider are stored in the cattle-globale-data ns. 
- From the auth provider, add 2 users -  User1, group1 to the downstream cluster as :
    - Cluster Owner
- From the auth provider, create a project p1 and add 2 users - User3 and User4 to the cluster as:
   - Project Member
- Verify crtbs and prtbs created for each of these members in the local cluster
- Login as cluster owner and verify the cluster owner can perform the actions:
- Verified group membership
- Verified Nested group membership
- Verified multiple access modes
     - Set the auth access mode to Allow members of clusters and projects, plus authorized users & groups
         - Login as User5 and verify the access is not allowed.
         - Login as User1 and User3 and verify access is allowed.
    - Set the auth access mode to Restrict access to only the authorized users & groups 
         - Add the user1 and user2 to the list of allowed users in auth page
         - Login as user1 and user2 and verify the access is allowed.
        - Login as admin user and verify the user access is permitted. 
- Auth upgrade checks
- Enable, disable and re-enable auth providers
- Refresh user and group membership
- Auth user search
- Enable auth using terraform
-  Update auth config from CLI

**Comment by anupama2501** [comment]
Waiting on access to verify China end points.

**Comment by anupama2501** [comment]
Unable to validate china endpoints at this time. Closing the issue and will open a new ticket if we find any issues around it later.
